File: backend/schemas/book.py
"""
Book related schemas
"""
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field
from enum import Enum


class CategoryEnum(str, Enum):
    BUSINESS = "business"
    PSYCHOLOGY = "psychology"
    FICTION = "fiction"
    SCIENCE = "science"
    HISTORY = "history"
    PHILOSOPHY = "philosophy"


class SortEnum(str, Enum):
    POPULAR = "popular"
    NEWEST = "newest"
    MOST_DISCUSSED = "most_discussed"


class PeriodEnum(str, Enum):
    TODAY = "today"
    WEEK = "week"
    MONTH = "month"
    ALL = "all"


class BookTypeEnum(str, Enum):
    AI_KNOWN = "ai_known"
    VECTORIZED = "vectorized"


class Character(BaseModel):
    """Character schema"""
    id: str
    name: str
    alias: List[str] = []
    description: str
    personality: Optional[str] = None
    dialogue_count: int = 0
    enabled: bool = True

    class Config:
        from_attributes = True


class Book(BaseModel):
    """Basic book schema (for list views) matching API spec"""
    id: str
    title: str
    author: str
    cover: Optional[str] = None
    category: Optional[str] = None
    description: Optional[str] = None
    dialogue_count: int = 0
    rating: float = Field(default=0.0, ge=0, le=5)
    created_at: datetime

    class Config:
        from_attributes = True


class BookDetail(Book):
    """Detailed book schema matching API spec"""
    type: BookTypeEnum
    chapters: Optional[int] = None
    estimated_reading_time: Optional[int] = None  # in minutes
    characters: List[Character] = []
    tags: List[str] = []
    uploader: Optional[Dict[str, str]] = None  # {id, nickname}

    class Config:
        from_attributes = True


class BookList(BaseModel):
    """Book list response with pagination"""
    books: List[Book]
    pagination: Dict[str, Any]

    class Config:
        from_attributes = True


# Aliases for backwards compatibility
BookResponse = Book
CharacterResponse = Character
BookCreate = BaseModel
BookUpdate = BaseModel

File: frontend/src/types/book.ts
/**
 * Book-related type definitions
 * Based on API specification from .futurxlab/api-specification.yaml
 * Maintains business logic conservation between API and frontend
 */

/**
 * Book category enum matching API specification
 */
export enum BookCategory {
  BUSINESS = 'business',
  PSYCHOLOGY = 'psychology',
  FICTION = 'fiction',
  SCIENCE = 'science',
  HISTORY = 'history',
  PHILOSOPHY = 'philosophy'
}

/**
 * Book status enum
 */
export enum BookStatus {
  PUBLISHED = 'published',
  DRAFT = 'draft',
  REVIEW = 'review',
  PROCESSING = 'processing'
}

/**
 * Book difficulty levels
 */
export enum BookDifficulty {
  BEGINNER = 'beginner',
  INTERMEDIATE = 'intermediate',
  ADVANCED = 'advanced'
}

/**
 * Upload source types
 */
export enum UploadSource {
  USER = 'user',
  ADMIN = 'admin',
  SYSTEM = 'system'
}

/**
 * Sort options for book listings
 */
export enum BookSortOption {
  POPULAR = 'popular',
  RECENT = 'recent',
  RATING = 'rating',
  TITLE = 'title'
}

/**
 * Book character type for dialogue
 */
export interface BookCharacter {
  id: string;
  name: string;
  description: string;
  avatar?: string;
  personality?: string;
  dialogueStyle?: string;
  isMainCharacter: boolean;
  isAvailable: boolean;
}

/**
 * Core book type matching API schema
 */
export interface Book {
  id: string;
  title: string;
  author: string;
  isbn?: string;
  coverUrl?: string;
  description: string;
  category: BookCategory;
  difficulty: BookDifficulty;
  language: string;
  publishDate?: string;
  pageCount?: number;
  publisher?: string;
  rating: number;
  ratingCount: number;
  dialogueCount: number;
  uploadSource: UploadSource;
  uploadedBy?: string;
  status: BookStatus;
  vectorized: boolean;
  vectorizationProgress?: number;
  characters?: BookCharacter[];
  tags?: string[];
  createdAt: string;
  updatedAt: string;
}

/**
 * Book list response with pagination
 */
export interface BookListResponse {
  books: Book[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    total_pages: number;
    has_next: boolean;
    has_prev: boolean;
  };
}

/**
 * Book search result
 */
export interface BookSearchResult {
  book: Book;
  relevanceScore: number;
  matchedQuery?: string;
  snippet?: string;
}

/**
 * Search response
 */
export interface SearchResponse {
  results: BookSearchResult[];
  total: number;
  query: string;
  suggestions?: string[];
}

/**
 * Book filter parameters
 */
export interface BookFilters {
  category?: BookCategory;
  difficulty?: BookDifficulty;
  language?: string;
  minRating?: number;
  hasCharacters?: boolean;
  isVectorized?: boolean;
}

/**
 * Book query parameters for API requests
 */
export interface BookQueryParams {
  page?: number;
  pageSize?: number;
  sort?: BookSortOption;
  filters?: BookFilters;
  search?: string;
}

/**
 * Popular books period options
 */
export enum PopularPeriod {
  TODAY = 'today',
  WEEK = 'week',
  MONTH = 'month',
  ALL = 'all'
}

/**
 * Book statistics for display
 */
export interface BookStats {
  totalBooks: number;
  totalDialogues: number;
  activeUsers: number;
  averageRating: number;
}

/**
 * Book recommendation based on user behavior
 */
export interface BookRecommendation {
  book: Book;
  reason: string;
  score: number;
}

/**
 * User book interaction tracking
 */
export interface UserBookInteraction {
  bookId: string;
  lastViewed: string;
  viewCount: number;
  hasDialogue: boolean;
  isFavorite: boolean;
}